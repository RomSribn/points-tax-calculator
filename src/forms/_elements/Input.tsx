import React, { DetailedHTMLProps, InputHTMLAttributes, forwardRef } from 'react';
import classNames from 'classnames';

export type InputSize = 'medium' | 'large';
export type InputType = 'text' | 'password' | 'email';

export type InputProps = {
  id: string;
  name: string;
  label: string;
  type?: InputType;
  className?: string;
} & Omit<DetailedHTMLProps<InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>, 'size'>;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const Input: any = forwardRef<HTMLInputElement, InputProps>(
  ({ id, name, label, type = 'text', className = '', placeholder, ...props }, ref) => {
    return (
      <div className="form-group">
        <input
          id={id}
          ref={ref}
          name={name}
          type={type}
          aria-label={label}
          placeholder={placeholder}
          className={classNames(['form-control', className])}
          {...props}
        />
      </div>
    );
  }
);

export { Input };
