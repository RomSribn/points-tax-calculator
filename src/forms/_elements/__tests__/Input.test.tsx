import React from 'react';
import { render, screen } from '@utils/tests';
import userEvent from '@testing-library/user-event';

import { Input } from '../Input';

describe('Input', () => {
  test('Should renders input element with correct attributes', () => {
    const id = 'input-id';
    const name = 'input-name';
    const label = 'Input Label';
    const placeholder = 'Enter a value';

    render(
      <Input
        id={id}
        name={name}
        label={label}
        placeholder={placeholder}
        className="custom-class"
        data-testid="input-element"
      />
    );

    const inputElement = screen.getByTestId('input-element') as HTMLInputElement;

    expect(inputElement).toBeInTheDocument();
    expect(inputElement.id).toBe(id);
    expect(inputElement.name).toBe(name);
    expect(inputElement.type).toBe('text');
    expect(inputElement.getAttribute('aria-label')).toBe(label);
    expect(inputElement.placeholder).toBe(placeholder);
    expect(inputElement.className).toContain('form-control');
    expect(inputElement.className).toContain('custom-class');

    expect(inputElement).toMatchSnapshot();
  });

  test('Should handles input changes', () => {
    const handleChange = jest.fn();
    const inputValue = 'Test Value';

    render(<Input id="input-id" name="input-name" label="Input Label" onChange={handleChange} />);

    const inputElement = screen.getByLabelText('Input Label') as HTMLInputElement;

    userEvent.type(inputElement, inputValue);

    expect(handleChange).toHaveBeenCalledTimes(inputValue.length);
    expect(inputElement.value).toBe(inputValue);
  });
});
