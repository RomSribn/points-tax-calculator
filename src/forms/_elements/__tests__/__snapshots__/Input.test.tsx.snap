// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Should renders input element with correct attributes 1`] = `
<input
  aria-label="Input Label"
  class="form-control custom-class"
  data-testid="input-element"
  id="input-id"
  name="input-name"
  placeholder="Enter a value"
  type="text"
/>
`;
